router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala

```diff
@@ -11,40 +14,71 @@ object ClassifierFilter {
-    case ReqRep(req, Return(rep: Response)) if rep.headers.contains(SuccessClassHeader) =>
-      val success = rep.headers.get(SuccessClassHeader)
-        .flatMap { value =>
-          Try(value.toDouble).toOption
-        }.getOrElse(0.0)
-      if (success > 0.0)
-        ResponseClass.Successful(success)
-      else
-        ResponseClass.Failed(false)
+  private[this] object ResponseSuccessClass {
+    @inline def unapply(headers: Headers): Option[ResponseClass] =
+      headers.get(SuccessClassHeader).map { value =>
+        val success = value.toDouble
+        if (success > 0.0) ResponseClass.Successful(success)
+        else ResponseClass.Failed(false)
+      }
   }
-}
 
-class ClassifierFilter(classifier: ResponseClassifier) extends SimpleFilter[Request, Response] {
+  object SuccessClassClassifier extends H2Classifier {
+    override val streamClassifier: PartialFunction[H2ReqRepFrame, ResponseClass] = {
+      case H2ReqRepFrame(_, Return((_, Some(Return(ResponseSuccessClass(c)))))) => c
+    }
 
-  private[this] val successClass = classifier.andThen(_.fractionalSuccess).lift
+    override val responseClassifier: PartialFunction[H2ReqRep, ResponseClass] = {
+      case H2ReqRep(_, Return(rsp)) if rsp.headers.contains(SuccessClassHeader) =>
+        rsp.headers.get(SuccessClassHeader).map { value =>
+          val success = value.toDouble
+          if (success > 0.0) ResponseClass.Successful(success)
+          else ResponseClass.Failed(false)
+        }.get
+    }
+  }
+}
 
+class ClassifierFilter(classifier: H2Classifier) extends SimpleFilter[Request, Response] {
+  import ClassifierFilter.SuccessClassHeader
+  private[this] val successHeader: ResponseClass => String =
+    _.fractionalSuccess.toString
   def apply(req: Request, svc: Service[Request, Response]): Future[Response] = {
-    svc(req).map { rep =>
-      successClass(ReqRep(req, Return(rep))) match {
-        case Some(success) =>
-          rep.headers.set(ClassifierFilter.SuccessClassHeader, success.toString)
-        case None =>
+    svc(req).map { rep: Response =>
+      if (rep.stream.isEmpty) {
+        // classify early - response class goes in headers
+        val success =
+          classifier.responseClassifier
+            .lift(H2ReqRep(req, Return(rep)))
+            // the responseClassifier is non-total, it won't classify successes
+            .getOrElse(ResponseClass.Success)
+        rep.headers.set(SuccessClassHeader, successHeader(success))
+        rep
+      } else {
+        // if the early classification attempt is not defined, attempt
+        // late classification on the last frame in the response stream
+        val stream = rep.stream.onFrame {
+          case Return(frame: Trailers) =>
+            val success =
+              classifier
+                .streamClassifier(H2ReqRepFrame(req, Return(rep), Some(Return(frame))))
+            frame.set(SuccessClassHeader, successHeader(success))
+          case _ =>
```

Comment:
in this case we want to add a Trailers frame to the stream.  To do this we probably need something like

```
class StreamFlatMap(underlying: Stream, f: Frame => Seq[Frame]) extends Stream { 
  val q = new QueueOfSomeKind
  def read(): Future[Frame] = {
    if (q.nonEmpty) Future.value(q.pop)
    else underlying.read().map(f).map { fs => q.push(fs) ; q.pop }
  }
}
```